### DISCUSSION TITLE
How to initialize objects in collision free random states?

### DISCUSSION THREAD

**Discussion by Ankur-Deka** [google-deepmind/mujoco#47] [Asking for Help] [answered]
The current approach I am using is to initialize in random states and then check for collision using a script which is similar to the following.

```
'''
code to initialize in random state
'''

try:
          self.sim.forward()
except mujoco_py.MujocoException as e:
          continue
contacts = self.get_contacts(object)
if len(contacts) > 0:
          print('collision detected, try sampling a new state')
```

It works most of the time but once in a while it fails. The object is initialized in a collision state and gets thrown away in the scene and the snippet above doesn't detect the collision. Since I am calling sim.forward (and NOT sim.step), [there shouldn't be any integration in time as per documentation](https://github.com/deepmind/mujoco/blob/1f7eaae62e0cd45a71cf7a593fa5605720765167/include/mujoco.h#L154) and correspondingly the plate shouldn't have been thrown away. How can I avoid the object being thrown away or how can I correctly sample collision free random states using an alternate approach?

**Comment by erez-tom** [comment, marked as answer]
Your approach is correct, so this seems to me like a `mujoco-py` issue.  One idea to work around it could be to look at [efc_force](https://github.com/deepmind/mujoco/blob/main/include/mjdata.h#L305) or [qfrc_constraint](https://github.com/deepmind/mujoco/blob/main/include/mjdata.h#L307) -- these should be 0 if and only if there are no contacts (and your model doesn't use constraints).
